src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -636,11 +685,39 @@ public void OnReset(Exception resetException, bool canRetry = false)
                         canRetry = false;
                     }
 
-                    _resetException = resetException;
-                    _canRetry = canRetry;
+                    // Per section 8.1 in the RFC:
+                    // If the server has completed the response body (i.e. we've received EndStream)
+                    // but the request body is still sending, and we then receive a RST_STREAM with errorCode = NO_ERROR,
+                    // we treat this specially and simply cancel sending the request body, rather than treating
+                    // the entire request as failed.
+                    if (resetStreamErrorCode == Http2ProtocolErrorCode.NoError && 
+                        _responseCompletionState == StreamCompletionState.Completed)
+                    {
+                        if (_requestCompletionState == StreamCompletionState.InProgress)
+                        {
+                            _requestBodyAbandoned = true;
+                            requestBodyCancellationSource = _requestBodyCancellationSource;
+                            Debug.Assert(requestBodyCancellationSource != null);
```

Comment:
Up to you, but would it be simpler to just make this:
```C#
_requestBodyAbandoned = true;
_requestBodyCancellationSource.Cancel();
return;
```
and then delete the if/else below and just have it Cancel?

Reply:
I'm reluctant to call CancellationTokenSource.Cancel while holding the lock here. This can call arbitrary code.

If you think I'm being too conservative, let me know.

Reply:
> I'm reluctant to call CancellationTokenSource.Cancel while holding the lock here.

Ah, that makes sense.  The diff wasn't highlighting that we were in a lock.  If it's important that we're not holding the lock, then, consider adding a Debug.Assert(!Monitor.IsEntered(...)) at the top of the function.